modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteTableModify.java

```diff
@@ -96,4 +113,85 @@ public IgniteTableModify(RelInput input) {
+                                                                         List<RelTraitSet> inTraits) {
+        IgniteDistribution distributionTrait = IgniteDistributions.affinity(0, "default",
+                nodeTraits.getTrait(DistributionTraitDef.INSTANCE));
+
+        IgniteDistribution targetDistributionTrait = IgniteDistributions.affinity(0, "default",
+                table.getDistribution());
+
+        if ((distributionTrait.function().type() != RelDistribution.Type.HASH_DISTRIBUTED) ||
+                !(DistributionFunction.satisfy(distributionTrait.function(), targetDistributionTrait.function()))) {
+            return Collections.emptyList();
+        }
+
+        return ImmutableList.of(Pair.of(nodeTraits.replace(TraitUtils.distribution(inTraits.get(0))),
+                inTraits));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override public List<Pair<RelTraitSet, List<RelTraitSet>>> deriveCollation(RelTraitSet nodeTraits,
+                                                                      List<RelTraitSet> inTraits) {
+        return ImmutableList.of(Pair.of(nodeTraits.replace(TraitUtils.collation(inTraits.get(0))),
+                inTraits));
+    }
+
+    /**
+     *
+     */
+    @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughCorrelation(RelTraitSet nodeTraits,
+                                                                       List<RelTraitSet> inTraits) {
+        CorrelationTrait correlation = TraitUtils.correlation(nodeTraits);
+
+        return Pair.of(nodeTraits, ImmutableList.of(inTraits.get(0).replace(correlation)));
+    }
+
+    /**
+     *
+     */
+    @Override public List<Pair<RelTraitSet, List<RelTraitSet>>> deriveCorrelation(RelTraitSet nodeTraits,
+                                                                        List<RelTraitSet> inTraits) {
+        return ImmutableList.of(Pair.of(nodeTraits.replace(TraitUtils.correlation(inTraits.get(0))), inTraits));
+    }
+
+    /** {@inheritDoc} */
+    @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+        IgniteCostFactory costFactory = (IgniteCostFactory)planner.getCostFactory();
+
+        double rowCnt = mq.getRowCount(getInput());
+        double bytesPerRow = getRowType().getFieldCount() * IgniteCost.AVERAGE_FIELD_SIZE;
+        double totalBytes = rowCnt * bytesPerRow;
+        double cpuCost;
+        double networkCost = 0;
+
+        cpuCost = rowCnt * IgniteCost.ROW_PASS_THROUGH_COST;
+
+        IgniteDistribution distributionTrait = IgniteDistributions.affinity(0, "default",
+                getTraitSet().getTrait(DistributionTraitDef.INSTANCE));
+
+        if (distributionTrait.function().type() == RelDistribution.Type.HASH_DISTRIBUTED &&
+                distributionTrait.function().affinity()) {
+            networkCost = 0;
+        } else {
+            networkCost = totalBytes;
+        }
```

Comment:
this piece if code does't make any sense

Reply:
The idea is to express that if the distribution is hash and the cache has affinity, then there is no network cost since there will be no need to move around data. Could you please let me know what needs to be corrected here?